File: bumperbot_vision/bumperbot_vision/stereo_vision.py
#!/usr/bin/env python3
import numpy as np
import cv2
from cv_bridge import CvBridge

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image, CameraInfo, PointCloud2, PointField
import sensor_msgs_py.point_cloud2 as pc2

import message_filters

class StereoVision(Node):
    def __init__(self, name):
        super().__init__(name)
        
        self.bridge = CvBridge()

        self.left_image_sub = message_filters.Subscriber(
            self, Image, "left_camera/image_raw", qos_profile=qos_profile_sensor_data)
        
        self.left_info_sub = message_filters.Subscriber(
            self, CameraInfo, "left_camera/camera_info", qos_profile=qos_profile_sensor_data)
        
        self.right_image_sub = message_filters.Subscriber(
            self, Image, "right_camera/image_raw", qos_profile=qos_profile_sensor_data)
        
        self.right_info_sub = message_filters.Subscriber(
            self, CameraInfo, "right_camera/camera_info", qos_profile=qos_profile_sensor_data)

        self.time_sync = message_filters.ApproximateTimeSynchronizer(
            [self.left_image_sub, self.left_info_sub, self.right_image_sub, self.right_info_sub],
            queue_size=10,
            slop=0.1
        )
        self.time_sync.registerCallback(self.image_callback)

        self.disparity_image_pub = self.create_publisher(
            Image, "disparity", qos_profile_sensor_data)
        
        self.point_cloud_pub = self.create_publisher(
            PointCloud2, "point_cloud", qos_profile_sensor_data)

        self.get_logger().info("Stereo Vision Node Initialized")

    def image_callback(self, left_img_msg, left_info_msg, right_img_msg, right_info_msg):
        """
        Callback triggered only when all 4 messages are available and synchronized.
        """
        try:
            left_img = self.bridge.imgmsg_to_cv2(left_img_msg, desired_encoding="mono8")
            right_img = self.bridge.imgmsg_to_cv2(right_img_msg, desired_encoding="mono8")

            sgbm = cv2.StereoSGBM_create(
                minDisparity=0,
                numDisparities=96,
                blockSize=9,
                P1=8 * 9 * 9,
                P2=32 * 9 * 9,
                disp12MaxDiff=1,
                preFilterCap=63,
                uniquenessRatio=10,
                speckleWindowSize=100,
                speckleRange=32,
                mode=cv2.STEREO_SGBM_MODE_SGBM
            )
            disparity_sgbm = sgbm.compute(left_img, right_img)

            disparity = disparity_sgbm.astype(np.float32) / 16.0
            normalized_disparity = cv2.normalize(
                disparity_sgbm, None, alpha=0, beta=255, 
                norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_8U
            )
            disparity_out_msg = self.bridge.cv2_to_imgmsg(normalized_disparity, encoding="mono8")
            disparity_out_msg.header = left_img_msg.header
            self.disparity_image_pub.publish(disparity_out_msg)

            fx = left_info_msg.k[0]
            cx = left_info_msg.k[2]
            fy = left_info_msg.k[4]
            cy = left_info_msg.k[5]
            
            baseline = 0.06 
            mask = (disparity > 10.0) & (disparity < 96.0)

            if not np.any(mask):
                return

            rows, cols = left_img.shape
            u, v = np.meshgrid(np.arange(cols), np.arange(rows))
            
            u_valid = u[mask]
            v_valid = v[mask]
            d_valid = disparity[mask]
            intensity_valid = left_img[mask]

            z = (fx * baseline) / d_valid
            x = (u_valid - cx) * z / fx
            y = (v_valid - cy) * z / fy
            intensity = intensity_valid.astype(np.float32) / 255.0
            points = np.stack([x, y, z, intensity], axis=-1)

            fields = [
                PointField(name='x', offset=0, datatype=PointField.FLOAT32, count=1),
                PointField(name='y', offset=4, datatype=PointField.FLOAT32, count=1),
                PointField(name='z', offset=8, datatype=PointField.FLOAT32, count=1),
                PointField(name='intensity', offset=12, datatype=PointField.FLOAT32, count=1),
            ]

            pc_msg = pc2.create_cloud(left_img_msg.header, fields, points)
            self.point_cloud_pub.publish(pc_msg)

        except Exception as e:
            self.get_logger().error(f"Error in stereo callback: {e}")

def main(args=None):
    rclpy.init(args=args)
    node = StereoVision("stereo_vision")
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: bumperbot_vision/src/stereo_vision.cpp
#include "bumperbot_vision/stereo_vision.hpp"

#include "cv_bridge/cv_bridge.hpp"

#include "opencv2/core.hpp"
#include "opencv2/opencv.hpp"
#include "sensor_msgs/point_cloud2_iterator.hpp"

namespace bumperbot_vision
{
StereoVision::StereoVision(const std::string & name) : Node(name)
{
  approximate_sync_ = std::make_shared<ApproximateSync>(
    ApproximatePolicy(10), left_image_sub_, left_info_sub_, right_image_sub_, right_info_sub_);
  approximate_sync_->registerCallback(
    std::bind(&StereoVision::imageCallback, this, _1, _2, _3, _4));
  image_transport::TransportHints hints{this};
  const auto sensor_data_qos = rclcpp::SensorDataQoS().get_rmw_qos_profile();

  left_image_sub_.subscribe(
    this, "left_camera/image_raw", hints.getTransport(), sensor_data_qos);
  left_info_sub_.subscribe(this, "left_camera/camera_info", sensor_data_qos);
  right_image_sub_.subscribe(
    this, "right_camera/image_raw", hints.getTransport(), sensor_data_qos);
  right_info_sub_.subscribe(this, "right_camera/camera_info", sensor_data_qos);

  disparity_image_pub_ =
    image_transport::create_publisher(this, "disparity", rmw_qos_profile_sensor_data);
  point_cloud_pub_ =
    create_publisher<sensor_msgs::msg::PointCloud2>("point_cloud", rclcpp::SensorDataQoS());
}

void StereoVision::imageCallback(
  const sensor_msgs::msg::Image::ConstSharedPtr & left_image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & left_info_msg,
  const sensor_msgs::msg::Image::ConstSharedPtr & right_image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & right_info_msg)
{
  cv::Mat left = cv_bridge::toCvCopy(left_image_msg, left_image_msg->encoding)->image;
  cv::Mat right = cv_bridge::toCvCopy(right_image_msg, right_image_msg->encoding)->image;
  cv::Ptr<cv::StereoSGBM> sgbm =
    cv::StereoSGBM::create(0, 96, 9, 8 * 9 * 9, 32 * 9 * 9, 1, 63, 10, 100, 32);
  cv::Mat disparity_sgbm, disparity;
  sgbm->compute(left, right, disparity_sgbm);
  disparity_sgbm.convertTo(disparity, CV_32F, 1.0 / 16.0);
  cv::Mat normalized_disparity;
  double min_disp = 0.0;
  double max_disp = 96.0 * 16.0;
  cv::normalize(disparity_sgbm, normalized_disparity, 0, 255, cv::NORM_MINMAX, CV_8U);

  auto disparity_msg =
    cv_bridge::CvImage(std_msgs::msg::Header(), "mono8", normalized_disparity).toImageMsg();
  disparity_msg->header = left_image_msg->header;
  disparity_image_pub_.publish(disparity_msg);

  // Generate PointCloud
  sensor_msgs::msg::PointCloud2 pointcloud_msg;
  pointcloud_msg.header = left_image_msg->header;
  pointcloud_msg.height = 1;
  pointcloud_msg.width = left.rows * left.cols;
  pointcloud_msg.is_dense = false;
  sensor_msgs::PointCloud2Modifier modifier(pointcloud_msg);
  modifier.setPointCloud2Fields(
    4, "x", 1, sensor_msgs::msg::PointField::FLOAT32, "y", 1, sensor_msgs::msg::PointField::FLOAT32,
    "z", 1, sensor_msgs::msg::PointField::FLOAT32, "intensity", 1,
    sensor_msgs::msg::PointField::FLOAT32);

  sensor_msgs::PointCloud2Iterator<float> iter_x(pointcloud_msg, "x");
  sensor_msgs::PointCloud2Iterator<float> iter_y(pointcloud_msg, "y");
  sensor_msgs::PointCloud2Iterator<float> iter_z(pointcloud_msg, "z");
  sensor_msgs::PointCloud2Iterator<float> iter_intensity(pointcloud_msg, "intensity");

  for (int v = 0; v < left.rows; v++) {
    for (int u = 0; u < left.cols; u++) {
      if (disparity.at<float>(v, u) <= 10.0 || disparity.at<float>(v, u) >= 96.0) {
        continue;
      }
      double x = (u - left_info_msg->k[2]) / left_info_msg->k[0];
      double y = (v - left_info_msg->k[5]) / left_info_msg->k[4];
      double depth = left_info_msg->k[0] * 0.06 / (disparity.at<float>(v, u));
      *iter_x = x * depth;
      *iter_y = y * depth;
      *iter_z = depth;
      *iter_intensity = left.at<uchar>(v, u) / 255.0;
      ++iter_x;
      ++iter_y;
      ++iter_z;
      ++iter_intensity;
    }
  }
  point_cloud_pub_->publish(pointcloud_msg);
}
}  // namespace bumperbot_vision

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<bumperbot_vision::StereoVision>("stereo_vision"));
  rclcpp::shutdown();
  return 0;
}